### PR TITLE
Expression evaluator

### PR DESCRIPTION
**Pull request by VinInn**
A new Expression Evaluator based on the standard C++ compiler
to replace the current one that is implemented using on a home-made parser (based on boost::spirit) and access objects using ROOT-reflection

motivations, doc, examples in 
http://innocent.home.cern.ch/innocent/ExprEval.pdf
to be presented at next reco/phys-tool meeting

### CODE REVIEW

**Comment by degano** [comment]
@cmsbuild please test

**Comment by VinInn** [comment]
@cmsbuild ,
ok let's try again...

**Comment by degano** [comment]
@cmsbuild please test

**Comment by VinInn** [comment]
to avoid polluting the release with test packages I am trying to create and compile it on the fly with

```
   system("rm -rf $CMSSW_BASE/src/VITest $CMSSW_BASE/include/$SCRAM_ARCH/VITest");
   system("echo $CMSSW_BASE");
   system("mkdir -p $CMSSW_BASE/src/VITest/ExprEval/src; cp $CMSSW_BASE/src/CommonTools/Utils/test/ExprEvalStubs/*.h $CMSSW_BASE/src/VITest/ExprEval/src/.");
   system("cp $CMSSW_BASE/src/CommonTools/Utils/test/ExprEvalStubs/BuildFile.xml $CMSSW_BASE/src/VITest/ExprEval/.; pushd $CMSSW_BASE; scram b -j 8; popd");
   system("ls $CMSSW_BASE/src; ls -l $CMSSW_BASE/src/VITest/ExprEval/src");
   system("ls -l $CMSSW_BASE/include/$SCRAM_ARCH/VITest/ExprEval/src");
```

need to understand what goes wrong in jenkins build w/r/t/ my own area

**Comment by VinInn** [comment]
@cmsbuild please test

**Comment by degano** [comment]
@cmsbuild please test

**Comment by VinInn** [comment]
Wait, discovered a concurrency issue when running test in parallel....
fixing code right now...

**Comment by VinInn** [comment]
@cmsbuild please test

this time tests running in parallel should not step on each other toes

**Comment by degano** [comment]
@cmsbuild please test
Aborted the previous tests, restarting now.

**Comment by VinInn** [comment]
cannot be...
@cmsbuild please test
run unit tests
relvals CANNOT be affected by this PR

**Comment by slava77** [comment]
We need a bit better error detection/interpretation in the jenkins/cms-bot tests.
If the same workflow is broken in the reference IB, the error should at least have a note of that.
Otherwise it triggers too much noise that takes an effort to interpret for people who don't watch the whole cmssw repo.
@ktf @nclopezo @degano 
Please check or follow up.
Thank you.

**Comment by degano** [comment]
@slava77 @VinInn this is indeed a problem with the latest IB (2015-01-27-0200). 
I'll re-trigger the tests once the afternoon IB is available (where that issue has been fixed).

**Comment by davidlange6** [comment]
I agree - maybe some simple mask can be added that compares against the baseline (ok, this implies having a baseline though)

**Comment by davidlange6** [comment]
please test

**Comment by degano** [comment]
@cmsbuild please test

**Comment by VinInn** [comment]
doing scram b from inside each given test does not work, they clash
need to find a different approach (at least this time one of the two succeeded...)

**Comment by smuzaffar** [comment]
@VinInn , yes doing "scram build" within a "scram build" is not a right approach. I think right approach should be to move EEUnitTest/ExprEval/src/*.h files in to CommonTools/Utils/src and let scram compile then when you do first scram build.

**Comment by VinInn** [comment]
I was hoping NOT to pollute the release in particular with a huge gch used only for my tests.
In addition it may create circular dependencies...
what I will do is to create a subsystem ExprEvalTest with all the files already in there
we will then decide how to make it "appearing" during the tests if we do not want to have it around in the distribution...

**Comment by VinInn** [comment]
@cmsbuild please test

**Comment by VinInn** [comment]
Now the test package is in the "release"
I am sure we can find a way to hide it somewhere and materialize it only at scram b runtests time...

**Comment by smuzaffar** [comment]
let me check if we can do scram b with in scram b runtests

**Comment by smuzaffar** [comment]
I think if you do something like the following then it should work
- change CommonTools/Utils/test/BuildFile.xml to have
<bin file="ExpressionEvaluatorUnitTest.cpp">
  <flag TEST_RUNNER_CMD="src/CommonTools/Utils/test/ExpressionEvaluatorUnitTest.sh"/>
  <use name="CommonTools/Utils"/>
</bin>

- create CommonTools/Utils/test/ExpressionEvaluatorUnitTest.sh script (with execution permission) which should copy the $CMSSW_BASE/src/CommonTools/Utils/test/ExprEvalStubs contents in to  $CMSSW_BASE/src/EEUnitTest/ExprEval/src and then do "scram build EEUnitTest/ExprEval"

**Comment by ktf** [comment]
@cmsbuild please test

**Comment by degano** [comment]
Test comment

**Comment by VinInn** [comment]
the real problem is that I have two tests:
```
<bin   name="testExpressionEvaluator" file="testExpressionEvaluator.cc,testRunner.cpp">
  <flags TEST_RUNNER_CMD="src/CommonTools/Utils/test/ExpressionEvaluatorIntTest.sh"/>

<bin file="ExpressionEvaluatorUnitTest.cpp">
<flags TEST_RUNNER_CMD="src/CommonTools/Utils/test/ExpressionEvaluatorIntTest.sh"/>

```

with
```
cat CommonTools/Utils/test/ExpressionEvaluatorIntTest.sh
echo running ExpressionEvaluatorIntTest
mkdir -p $CMSSW_BASE/src/ExpressionEvaluatorTests
cp -r $CMSSW_BASE/src/CommonTools/Utils/test/ExpressionEvaluatorTests/EEIntTest $CMSSW_BASE/src/ExpressionEvaluatorTests/EEIntTest
scram b ExpressionEvaluatorTests/EEIntTest
ls -l $CMSSW_BASE/include/$SCRAM_ARCH/ExpressionEvaluatorTests/EEIntTest/src


[innocent@vinavx2 src]$ cat CommonTools/Utils/test/ExpressionEvaluatorUnitTest.sh 
echo running ExpressionEvaluatorUnitTest
mkdir -p $CMSSW_BASE/src/ExpressionEvaluatorTests
cp -r $CMSSW_BASE/src/CommonTools/Utils/test/ExpressionEvaluatorTests/EEUnitTest $CMSSW_BASE/src/ExpressionEvaluatorTests/EEUnitTest
scram b ExpressionEvaluatorTests/EEUnitTest
ls -l $CMSSW_BASE/include/$SCRAM_ARCH/ExpressionEvaluatorTests/EEUnitTest/src

```

the two "scram b" interferes 

is it possible to have just one script that runs ONCE?
what I need to do is
```
cp -r $CMSSW_BASE/src/CommonTools/Utils/test/ExpressionEvaluatorTests/ $CMSSW_BASE/src/ExpressionEvaluatorTests
scram b ExpressionEvaluatorTests
ls -l $CMSSW_BASE/include/$SCRAM_ARCH/ExpressionEvaluatorTests/*/src
$CMSSW_BASE/test/$SCRAM_ARCH/ExpressionEvaluatorUnitTest
$CMSSW_BASE/test/$SCRAM_ARCH/testExpressionEvaluator
rm -rf  $CMSSW_BASE/src/ExpressionEvaluatorTests
```

**Comment by smuzaffar** [comment]
I remember now that, while running runtests/unittests during IB 
validation, we disallow "scram b" otherwise it will rebuild every thing 
(remember we install release using RPM at the validation time).

**Comment by smuzaffar** [comment]
yes, it is possible to run only one tests. You need to disable one test 
by adding the NO_TESTRUN flag e.g.|
<bin name="testExpressionEvaluator" 
file="testExpressionEvaluator.cc,testRunner.cpp">|
   <flags NO_TESTRUN="1"/>
</bin>

but as I mentioned these tests will only run if they are part of release 
itself (scram build during IB tests will not compile any new test).

**Comment by VinInn** [comment]
ok, so way need to be build at scram b time
ok
now assume I have the new "subsystem" in
CommonTools/Utils/test/ExpressionEvaluatorTests
can this be build and produce pch under
$CMSSW_BASE/include/$SCRAM_ARCH/ExpressionEvaluatorTests
(even $CMSSW_BASE/include/$SCRAM_ARCH/CommonTools/Utils/test/ExpressionEvaluatorTests ..)
how we do it for classes_def.xml?

**Comment by smuzaffar** [comment]
currently pch  are only generated for Subsystem/Package/src/precompile.h 
files. I would suggest to put your precompiler.h in the 
CommonTools/Utils/src directory (along with the other headers needed by it).

**Comment by VinInn** [comment]
I would like to avoid possible undesired/circular dependency.
I think I can locate "precompile.h" in CandUtils or RecoUtils

File: CommonTools/Utils/src/ExpressionEvaluator.cc [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+  std::string factory = "factory" + m_name;
+
+  std::string source = std::string("#include ")+quote+ pch +quote+"\n";
+  source+="struct "+m_name+" final : public "+iname + "{\n";
+  source+=iexpr;
+  source+="\n};\n";
+
+
+  source += "extern " + quote+'C'+quote+' ' + std::string(iname) + "* "+factory+"() {\n";
+  source += "static "+m_name+" local;\n";
```
Comment (VinInn) [inline CommonTools/Utils/src/ExpressionEvaluator.cc]: add const...

File: CommonTools/Utils/src/ExpressionEvaluator.cc [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+  void * dl = dlopen(ofile.c_str(),RTLD_LAZY);
+  if (!dl) {
+     remove(m_name);
+     throw  cms::Exception("ExpressionEvaluator", std::string("compilation/linking failed\n") +  cpp + ss + "dlerror " + dlerror());
+    return;
+  }
+
+  m_expr = dlsym(dl,factory.c_str());
+
```
Comment (VinInn) [inline CommonTools/Utils/src/ExpressionEvaluator.cc]: remove here

**Comment by VinInn** [comment]
ooiops need rebase
ok let's close this one and reopen a new one

**Pull request closed without merging**